### PR TITLE
Implement Ranges Set Operations

### PR DESCRIPTION
**Pull request by CaseyCarter**
* `mergeable` concept

* algorithms: `ranges::includes`, `ranges::set_union`, `ranges::set_intersection`, `ranges::set_difference`, and `ranges::set_symmetric_difference`.

Partially addresses #39.

### CODE REVIEW

File: stl/inc/algorithm [hunk 1]
```diff
@@ -1554,8 +1554,9 @@ namespace ranges {
         template <input_range _Rng, weakly_incrementable _Out>
             requires indirectly_copyable<iterator_t<_Rng>, _Out>
         constexpr copy_result<borrowed_iterator_t<_Rng>, _Out> operator()(_Rng&& _Range, _Out _Result) const {
-            auto _First   = _RANGES begin(_Range);
```
Comment (miscco) [inline stl/inc/algorithm:1557]: I remember someone talking about unrelated changes and `ranges::reverse` Ok that one is not merged yet

Comment (CaseyCarter) [reply stl/inc/algorithm:1557]: These are fallout from wrapping to 120 - and they almost certainly conflict with the `copy_backward` PR. I'll revert.

Comment (CaseyCarter) [reply stl/inc/algorithm]: There: all green now.

File: tests/std/tests/P0896R4_ranges_alg_includes/test.cpp [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+    static constexpr auto add_one = [](const int x) { return x + 1; };
+
+    template <ranges::input_range Haystack, ranges::input_range Needle>
+    static constexpr void call() {
+        using ranges::includes, ranges::less;
+
+        { // Validate range overload
+            Haystack haystack{haystack_elements};
+            Needle needle{needle_elements};
+            const same_as<bool> auto result = includes(haystack, needle, ranges::less{}, get_first, add_one);
```
Comment (miscco) [inline tests/std/tests/P0896R4_ranges_alg_includes/test.cpp:29]: You learn new something every day. I assume this is to prevent implicit conversions to bool. 

Neat!

File: stl/inc/algorithm [hunk 3]
```diff
@@ -6849,6 +7165,95 @@ _FwdIt3 set_symmetric_difference(_ExPo&&, _FwdIt1 _First1, _FwdIt1 _Last1, _FwdI
...
+            requires mergeable<_It1, _It2, _Out, _Pr, _Pj1, _Pj2>
+        constexpr set_symmetric_difference_result<_It1, _It2, _Out> operator()(_It1 _First1, _Se1 _Last1, _It2 _First2,
+            _Se2 _Last2, _Out _Result, _Pr _Pred = {}, _Pj1 _Proj1 = {}, _Pj2 _Proj2 = {}) const {
+            _Adl_verify_range(_First1, _Last1);
+            _Adl_verify_range(_First2, _Last2);
+            auto _UResult = _Set_symmetric_difference_unchecked(_Get_unwrapped(_STD move(_First1)),
+                _Get_unwrapped(_STD move(_Last1)), _Get_unwrapped(_STD move(_First2)),
+                _Get_unwrapped(_STD move(_Last2)), _Get_unwrapped_unverified(_STD move(_Result)), _Pass_fn(_Pred),
+                _Pass_fn(_Proj1), _Pass_fn(_Proj2));
+            _Seek_wrapped(_First1, _STD move(_UResult.in1));
```
Comment (miscco) [inline stl/inc/algorithm:7192]: I am wondering whether this one would be helpful (If yes than it is future work. No change requested):

```cpp
template<class _It1, class _It2, class _Out, class _UIt1, class _UIt2, class _UOut>
in_in_out_result<_It1, _It2, _Out> _Seek_wrapped(_It1& _In1, _It2& _In2, _UOut& _Out, in_in_out_result<_UIt1, _UIt2, _UOut>&& _UResult);
```
Same for `in_out_result `and `in_in_result` which occur often enough that it makes sense

File: tests/std/tests/P0896R4_ranges_alg_set_difference/test.cpp [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+            const same_as<set_difference_result<iterator_t<R1>, O>> auto result = set_difference(range1.begin(),
+                range1.end(), range2.begin(), range2.end(), O{output}, ranges::less{}, get_first, get_second);
+            assert(result.in == range1.end());
+            assert(result.out.peek() == output + ranges::size(elements1));
+            assert(equal(span{output}.first<ranges::size(elements1)>(), elements1));
+        }
+    }
+};
+
+template <class Continuation>
```
Comment (miscco) [inline tests/std/tests/P0896R4_ranges_alg_set_difference/test.cpp:80]: That is three times you use those instantiators. Even if we cannot find more uses for them it would be beneficial to move them to range_algorithm_helpers

File: tests/std/tests/P0896R4_ranges_alg_includes/test.cpp (hunk 2 above)
Comment (CaseyCarter) [reply tests/std/tests/P0896R4_ranges_alg_includes/test.cpp:29]: Yeah, I would have been using _type-constraint_ s like this instead of `static_assert(same_as<decltype(/*...*/), // ...` all along but I thought MSVC hadn't implemented them yet.

Comment (CaseyCarter) [reply tests/std/tests/P0896R4_ranges_alg_includes/test.cpp:29]: > I thought MSVC hadn't implemented them yet.

Apparently this is new for 16.7; I have no idea how they managed to sneak it past me.

File: tests/std/tests/P0896R4_ranges_alg_includes/test.cpp [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+            assert(result);
+        }
+
+        { // Validate range overload, needle not found
+            Haystack haystack{haystack_elements};
+            Needle needle{bad_needle_elements};
+            const same_as<bool> auto result = includes(haystack, needle, ranges::less{}, get_first, add_one);
+            assert(!result);
+        }
+        { // Validate iterator overload, needle not found
```
[the remark this answers is not in the document]
Comment (CaseyCarter) [reply tests/std/tests/P0896R4_ranges_alg_includes/test.cpp:60]: Yes, agreed: it's pointless to test this again.

File: tests/std/tests/P0896R4_ranges_alg_set_difference/test.cpp (hunk 4 above)
Comment (CaseyCarter) [reply tests/std/tests/P0896R4_ranges_alg_set_difference/test.cpp:80]: These are algorithm-specific enough that I prefer that they live in the algorithm-specific test sources.

File: stl/inc/algorithm (hunk 3 above)
Comment (CaseyCarter) [reply stl/inc/algorithm:7192]: Probably so. I think we should add a work item to scan through the unwrapping and rewrapping to find opportunities for simplification after/in conjunction with the pass to unwrap result iterators with `_Get_unwrapped_n`.

File: stl/inc/algorithm [hunk 6]
```diff
@@ -6621,6 +6621,78 @@ _NODISCARD bool includes(_ExPo&&, _FwdIt1 _First1, _FwdIt1 _Last1, _FwdIt2 _Firs
...
+namespace ranges {
+    // VARIABLE ranges::includes
+    class _Includes_fn : private _Not_quite_object {
+    public:
+        using _Not_quite_object::_Not_quite_object;
+
+        template <input_iterator _It1, sentinel_for<_It1> _Se1, input_iterator _It2, sentinel_for<_It2> _Se2,
+            class _Pj1 = identity, class _Pj2 = identity,
+            indirect_strict_weak_order<projected<_It1, _Pj1>, projected<_It2, _Pj2>> _Pr = ranges::less>
+        constexpr bool operator()(_It1 _First1, _Se1 _Last1, _It2 _First2, _Se2 _Last2, _Pr _Pred = {},
```
[the remark this answers is not in the document]
Comment (CaseyCarter) [reply stl/inc/algorithm:6635]: We need a GitHub Action that posts "Audit `_NODISCARD` because I know you forgot to so much as consider it." into PRs whose author is @CaseyCarter.

File: tests/std/tests/P0896R4_ranges_alg_includes/test.cpp [hunk 7]
```diff
@@ -0,0 +1,104 @@
...
+#include <cassert>
+#include <concepts>
+#include <ranges>
+#include <utility>
+
+#include <range_algorithm_support.hpp>
+
+using namespace std;
+using P = pair<int, int>;
+
```
[the remark this answers is not in the document]
Comment (CaseyCarter) [reply tests/std/tests/P0896R4_ranges_alg_includes/test.cpp:14]: It accepts a range but doesn't return an iterator - which is unusual in the Ranges design, we typically return at least an end iterator for any range an algorithm necessarily traverses fully. Our belief was that algorithms that return a single derived value like `all_of` and `includes` would be rendered less usable by adding multiple returns, so we don't strictly follow Stepanov's Rule of Useful Return.

[the remark this answers is not in the document]
Comment (mnatsuhara) [reply tests/std/tests/P0896R4_ranges_alg_includes/test.cpp:14]: Interesting -- I hadn't thought of that! Thanks for the explanation :)

File: stl/inc/algorithm [hunk 8]
```diff
@@ -6784,6 +7020,87 @@ _FwdIt3 set_difference(_ExPo&& _Exec, _FwdIt1 _First1, _FwdIt1 _Last1, _FwdIt2 _
...
+                if (_STD invoke(_Pred, _STD invoke(_Proj1, *_First1), _STD invoke(_Proj2, *_First2))) {
+                    *_Result = *_First1;
+                    ++_Result;
+                    ++_First1;
+                } else {
+                    if (!_STD invoke(_Pred, _STD invoke(_Proj2, *_First2), _STD invoke(_Proj1, *_First1))) {
+                        ++_First1;
+                    }
+
+                    ++_First2;
```
[the remark this answers is not in the document]
Comment (CaseyCarter) [reply stl/inc/algorithm:7095]: Yes, that is the case. The only control paths that exit the `for` loop are `return` statements which certainly return a value ;)

[the remark this answers is not in the document]
Comment (mnatsuhara) [reply stl/inc/algorithm:7095]: That's so fancy :)  I love when compilers are smart! 🤓

**Comment by CaseyCarter** [comment]
I'm having trouble coming up with a witty comment here... perhaps a limerick? What rhymes with `set_symmetric_difference`?
